### ISSUE TITLE
Too Many Load Balancers

### ISSUE THREAD

**Issue by pumoica** [skupperproject/skupper#964]
Hello,

Our use case is we have a central cluster deployed in AWS EKS and we need to connect a number of edge clusters to it. These edge clusters could be deployed on-site and/or in other public clouds.

We've noticed that for every linked edge cluster skupper creates 2 load balancers in our AWS environment. Unfortunately, our use case (longer term) assumes hundreds, if not thousands, of edge clusters to be linked to the central cluster and having twice as much of load balancers is simply a no go (from cost perspective).

Are there any skupper setup options to minimize the number of necessary load balancers, ideally, down to just one?

**Comment by grs** [comment]
If they are edge sites, then you could set --ingress=none  on init and then you will not have any loadbalancers exposed.

**Comment by pumoica** [comment]
But that would now allow to link the edge cluster to the central one. Would it?

**Comment by grs** [comment]
Yes, the edge cluster could still connect to the central one, and services could be invoked in either direction (and indeed from edge to edge).

**Comment by pumoica** [comment]
Could you please clarify how exactly to accomplish that?
When I init skupper (with ingress=none), as you mentioned, yes no additional load balancers are created. I then issue a token on the central cluster.
But when I go to the edge cluster, initiate a process of linking it (to the central cluster), and check the link status a I get this message: 

```
Link link1 not active (Failed to redeem claim: Post "https://skupper.<my_k8s_namespace>:8081..."
```
Which makes sense given the link is not valid. Or did I get this completely wrong and should rather follow a different linking procedure?

**Comment by grs** [comment]
my_k8s_namespace here is the central cluster?

**Comment by pumoica** [comment]
yes, it is the name of a k8s namespace in the central cluster where I initialized skupper in.

**Comment by grs** [comment]
That sounds like the token created by the central cluster's skupper is not setting a hostname as would be expected. What options was the central skupper site initialised with?

**Comment by pumoica** [comment]
No options other than the ingress one.

`skupper init --ingress=none`

**Comment by grs** [comment]
Sorry, it is the *edge* clusters that should use --ingress=none. The central cluster needs an ingress. The default will be a loadbalancer, but you can use a Contour http proxy or an nginx ingress (the ingress controller needs to have ssl passthrough enabled though for this).

**Comment by pumoica** [comment]
Got it. I installed the ingress controller (based on nginx) with the ssl passthrough but ran into a couple of issues.

1. with `--ingress=nginx-ingress-v1` skupper still initializes itself in the central cluster without a hostname. Setting an option `--ingress-host` to a domain and then creating an A record with that domain in Route 53 (associated with the load balancer of the ingress controller) solved the problem.
2. After linking the edge cluster the check of the link status returns: `http: server gave HTTP response to HTTPS client`

Any thoughts on how to address the second problem?

**Comment by grs** [comment]
Does the ingress controller have the --enable-ssl-passthrough option specified?

For the loadbalancer option, you should not need a separate loadbalancer for each edge that connects, unless they must all connect to different namespaces.

**Comment by grs** [comment]
Run curl -v against the url from the token (the one mentioned in any error in skupper link status). You should see a skupper signed cert, not the nginx default, cert being offered.

**Comment by grs** [comment]
Sorry, you also need the -k option.

I think your are running curl on windows there? Not familiar with the expected output there. Try running the same curl command from the skupper-router pod in one of the edge sites, i.e. `kubectl exec -it  <skupper-router-xxxxxxx-xxxx> -- curl -k -v https://host-from-token:443`.

I'm not aware of an example for sll-passthrough with nginx. However just running curl usually gives useful information (at least on linux!). In particular we should see the subject of the server certificate offered. If passthrough is working, that will be CN=skupper.

**Comment by pumoica** [comment]
Logged into the skupper router pod and executed the command. Here is its output:

```
*   Trying <ip stripped>...
* TCP_NODELAY set
* Connected to <link url stripped> port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* error:1408F10B:SSL routines:ssl3_get_record:wrong version number
* Closing connection 0
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
```

**Comment by pumoica** [comment]
yes

**Comment by grs** [comment]
That looks as though nginx is serving HTTP on the HTTPS port?

**Comment by pumoica** [comment]
The nginx serves both http and https without any crossover (http -> https or vice versa). I intentionally disabled http and left https only (again made sure there is no crossover) but the issue persists.

From what I gather the same symptoms might show up if nginx cannot identify a matching host to determine where to forward traffic to.

Do you guys have a working example of setting up the nginx ingress+aws eks+skupper that I could just replicate (instead of scouting through nginx/aws official docs and/or the Internet)?

**Comment by grs** [comment]
No we don't have any of our own instructions for setting up nginx ingress.

Skupper creates the ingress resource with the necessary host rules in it. All it requires is that ssl-passthrough is enabled.

Do the ingress-controller logs give any clues?

**Comment by pumoica** [comment]
Somehow, the problem has been fixed (had to dance with some annotations for the nginx controller and one combo took of the problem). Will try to narrow down to the root cause and if it helps can share the full details of the nginx ingress + aws eks + skupper here.

Meanwhile, could you please clarify my previous question about the load balancer and namespaces? Let's consider 3 edge clusters: 2 use one namespace `space-a` and the third one `space-b`. 

Do I understand correctly that I would still need 2 load balancers - one for the namespace `space-a` and another one for `space-b`?

**Comment by grs** [comment]
The namespaces at the edge don't affect the number of loadbalancers the central hub needs. If on the central hub you are running skupper in a single namespace, and connecting all the edges to that same skupper instance, then you only need the two[*] loadbalancers for that site, however many edges you have. 

([*] And you can avoid one of those if required, though it slightly changes the linking as you need to use tokens of type 'cert').

**Comment by pumoica** [comment]
I should have clarified earlier that was not referring to the namespaces at the edges, The example assumed namespaces in the central cluster.

Right now I have one load balancer (running for the nginx ingress controller) and at least was able to successfully link my edge cluster through it. I also see the skupper ingress associated with that load balancer/controller includes 4 host-based rules:

```
claims.<my_k8s_namespace_in_central_cluster>.<my_domain> --> skupper:8081
console.<my_k8s_namespace_in_central_cluster>.<my_domain> --> skupper:8080
inter-router.<my_k8s_namespace_in_central_cluster>.<my_domain> --> skupper:55671
edge.<my_k8s_namespace_in_central_cluster>.<my_domain> --> skupper:45671
```
These all go through that single load balancer.
What would I need the second load balancer for?

Back to my example, if I need to another namespace (in my central cluster) to be linked to some edge cluster, do I need a separate load balancer for it?
